src/main/java/bio/terra/service/tabulardata/google/bigquery/BigQueryDatasetPdao.java

```diff
@@ -565,6 +629,64 @@ public void insertIntoDatasetTable(
                 Optional.ofNullable(transactId).map(UUID::toString).orElse(null))));
   }
 
+  private static final String insertNewFileIdsIntoDatasetTableTemplate =
+      """
+    /* Soft delete all existing rows */
+    INSERT INTO `<project>.<dataset>.<softDeleteTable>` (<transactIdColumn>, <rowIdColumn>, <flightIdColumn>, <deletedAtColumn>, <deletedByColumn>)
+    SELECT @transactId, <rowIdColumn>, @flightId, CURRENT_TIMESTAMP(), @deletedBy
+    FROM `<project>.<dataset>.<liveTable>`;
+
+    /* Insert rows with new values */
+    INSERT INTO `<project>.<dataset>.<targetTable>` (<transactIdColumn>, <rowIdColumn>, <columns:{c|<c.name>}; separator=",">)
+    SELECT @transactId,GENERATE_UUID(),<columns:{c|<if(c.fileOrDirRef)>
+      <if(c.arrayOf)>
+        ARRAY(SELECT f.<newIdColumn> FROM UNNEST(<c.name>) i JOIN `<project>.<dataset>.<fileIdMappingTable>` f ON i = f.<origIdColumn>)
+      <else>
+        (SELECT <newIdColumn> FROM `<project>.<dataset>.<fileIdMappingTable>` f WHERE f.<origIdColumn> = t.<c.name>)
+      <endif>
+    <else>
+      <c.name>
+    <endif>}; separator=",">
+    FROM `<project>.<dataset>.<liveTable>` t;
+      """;
```

Comment:
I think because this is a special usecase where we need to join with the new table.  It's pretty low level, which is where we'd normally see these types of one offs that get hidden in the service layer.  I suppose we could have put this into the merge machinery but that seemed like it could complicate things

Reply:
My main concern is around maintainability long term -- If we change how transactions or our soft delete set up works, we'll need to remember to also update this code.